src/dosext/dpmi/dnative/sigsegv.c

```diff
@@ -531,19 +567,30 @@ static void signal_sas_wa(void)
 }
 #endif
 
-void signative_pre_init(void)
-{
-#if SIGRETURN_WA
-  if (config.cpu_vm_dpmi == CPUVM_NATIVE)
-    iret_frame_alloc();
-#endif
-}
-
 /* init_handler puts the handler in a sane state that glibc
    expects. That means restoring fs, gs and eflags for DPMI. */
 SIG_PROTO_PFX
 static void __init_handler(sigcontext_t *scp, unsigned long uc_flags)
 {
+#if SIGRETURN_WA
```

Comment:
the idea is to skip all the segment stuff if the signal comes from dosemu itself, then there's no need for it, just like without native DPMI.

But I need to check the SIGRETURN_WA, I'm not 100% sure precise bit is correct.

Reply:
No, no, please don't do that!
Esp in a patch that is marked entirely
differently. This hunk has nothing to
do with init/pre_init().
And its plain wrong, for example there
is this comment:
```
   * FIRST thing to do in signal handlers - to avoid being trapped into int0x11
   * forever, we must restore the eflags.
```
Not sure about alignment exception, but
I'd certainly restore at least DF before
running any host code.
Also it seems DS and ES are restored
later - even w/o that change. Same for
FS and GS.
Well if DS and ES are really still not restored
by the kernel, then this entire code is
unsafe. But I hope DS and ES are good,
FS and GS not. So I think the correct
change would be to double-check about
DS/ES and remove their restoring if no
longer needed. After that we need to make
sure all functions that are called before
restoring FS/GS are marked with SIG_PROTO_PFX,
which is currently not the case.

Reply:
The code is certainly tricky, especially for async signals if they trigger at some in between state. Best to be as careful as possible, particularly with the alignment flag.

The kernel resets DF on sighandler entry, that's an ABI requirement for function entry, and GCC-generated code depends on it (https://github.com/torvalds/linux/blob/9c1bec9c0b08abeac72ed6214b723adc224013bf/arch/x86/kernel/signal.c#L266)

DS and ES are set on signal entry for i386 but not x86_64 and are useless on x86_64 anyway (if you load ds and es in 64 bit mode it will look at the segment descriptor, but the base and limit from there are never used). I think you could actually keep DS/ES in registers for the duration of the signal handler, no need to even save it to the signal stack (but then you need to read DS/ES when you need it, that's annoying, and different from i386).

SS is subtle: https://github.com/torvalds/linux/blob/9c1bec9c0b08abeac72ed6214b723adc224013bf/arch/x86/kernel/signal_64.c#L216

and then we have fs and gs with the fsbase and gsbase semantics and convention, important to restore those ASAP.

Reply:
Thanks for clarifying.
I think it would make sense to restore
all segregs before checking for whether
the fault is from dosemu or not, and add
the comment that ds/es are irrelevant,
we restore them for consistency.

Reply:
in the end with #1960 in there was no need to move the order of anything here.